Mathlib/Data/Finsupp/Basic.lean

```diff
@@ -1058,78 +1058,84 @@ theorem mem_support_finset_sum [AddCommMonoid M] {s : Finset ι} {h : ι → α
-      split_ifs <;> simp [single_apply, *]
-    rw [Finsupp.curry, sum_apply, sum_apply, sum_eq_single, this, if_pos rfl]
-    · intro b _ b_ne
-      rw [this b, if_neg b_ne]
-    · intro _
-      rw [single_zero, single_zero, coe_zero, Pi.zero_apply, coe_zero, Pi.zero_apply]
+protected theorem uncurry_apply_pair (f : α →₀ β →₀ M) (a : α) (b : β) :
+    f.uncurry (a, b) = f a b :=
+  rfl
 
 @[simp]
-lemma curry_single (a : α × β) (m : M) :
-    (single a m).curry = single a.1 (single a.2 m) := by
-  simp [Finsupp.curry]
+lemma uncurry_single (a : α) (b : β) (m : M) :
+    (single a (single b m)).uncurry = single (a, b) m := by
+  ext ⟨x, y⟩
+  rcases eq_or_ne a x with rfl | hne <;> classical simp [single_apply, *]
 
-theorem sum_curry_index (f : α × β →₀ M) (g : α → β → M → N) (hg₀ : ∀ a b, g a b 0 = 0)
-    (hg₁ : ∀ a b c₀ c₁, g a b (c₀ + c₁) = g a b c₀ + g a b c₁) :
-    (f.curry.sum fun a f => f.sum (g a)) = f.sum fun p c => g p.1 p.2 c := by
-  rw [Finsupp.curry]
-  trans
-  · exact
-      sum_sum_index (fun a => sum_zero_index) fun a b₀ b₁ =>
-        sum_add_index' (fun a => hg₀ _ _) fun c d₀ d₁ => hg₁ _ _ _ _
-  congr; funext p c
-  trans
-  · exact sum_single_index sum_zero_index
-  exact sum_single_index (hg₀ _ _)
+theorem sum_uncurry_index [AddCommMonoid N] (f : α →₀ β →₀ M) (g : α × β → M → N) :
+    f.uncurry.sum (fun p c => g p c) = f.sum fun a f => f.sum fun b ↦ g (a, b) := by
+  simp only [Finsupp.sum, Finsupp.uncurry, Finset.sum_disjiUnion]
+  simp
 
-/-- Given a finitely supported function `f` from `α` to the type of
-finitely supported functions from `β` to `M`,
-`uncurry f` is the "uncurried" finitely supported function from `α × β` to `M`. -/
-protected def uncurry (f : α →₀ β →₀ M) : α × β →₀ M :=
-  f.sum fun a g => g.sum fun b c => single (a, b) c
+theorem sum_uncurry_index' [AddCommMonoid N] (f : α →₀ β →₀ M) (g : α → β → M → N) :
+    f.uncurry.sum (fun p c => g p.1 p.2 c) = f.sum fun a f => f.sum (g a) :=
+  sum_uncurry_index ..
 
-@[simp]
-protected theorem uncurry_apply_pair (f : α →₀ β →₀ M) (x : α) (y : β) :
-    f.uncurry (x, y) = f x y := by
-  rw [← curry_apply (f.uncurry) x y]
-  simp only [Finsupp.curry, Finsupp.uncurry, sum_sum_index, single_zero, single_add,
-    forall_true_iff, sum_single_index, single_zero, ← single_sum, sum_single]
+end Uncurry
+
+section Curry
 
-protected theorem uncurry_apply (f : α →₀ β →₀ M) (x : α × β) :
-    f.uncurry x = f x.1 x.2 := by
-  rw [← Prod.mk.eta (p := x), Finsupp.uncurry_apply_pair]
+variable [DecidableEq α] [Zero M]
+
+/-- Given a finitely supported function `f` from a product type `α × β` to `γ`,
+`curry f` is the "curried" finitely supported function from `α` to the type of
+finitely supported functions from `β` to `γ`. -/
+protected def curry (f : α × β →₀ M) : α →₀ β →₀ M where
+  toFun a :=
+    { toFun b := f (a, b)
+      support := f.support.filterMap (fun x ↦ if x.1 = a then x.2 else none) <| by simp +contextual
```

Comment:
Another option to avoid a decidable cascade would be to drop the new instance argument and construct the support classically, at least for now. 
```suggestion
      support := f.support.filterMap (fun x ↦ open scoped Classical in if x.1 = a then x.2 else none) <| by simp +contextual
```

Reply:
(it was already non computable before this PR anyway, and "make it computable" sounds better as a separate PR)

Reply:
I've already fixed compile by adding the required `DecidableEq` assumptions.